src/Components/Server/src/Circuits/ServerComponentDeserializer.cs

```diff
@@ -25,8 +25,12 @@ namespace Microsoft.AspNetCore.Components.Server
     // 'sequence' indicates the order in which this component got rendered on the server.
     // 'assemblyName' the assembly name for the rendered component.
     // 'type' the full type name for the rendered component.
+    // 'parameterDefinitions' a JSON serialized array that contains the definitions for the parameters including their names and types and assemblies.
```

Comment:
I want to understand why we'd need those things.

Reply:
Arguably the original code here is more correct, in that technically it's possible to supply arbitrary parameters to components, even when they don't declare any `[Parameter]` that can be reflected on from outside.

However I don't see any clear use case for that when it comes to setting up a root element, so would still prefer not to be round-tripping this info. I think it's a perfectly fine limitation to say that you can only supply parameter values for things with `[Parameter]` for this API.

Reply:
OK, I looked into changing the implementation to get the parameter types from the component type, and now see why the original code in this PR is the only reasonable way to do it. Basically it's as I mentioned before ("technically it's possible to supply arbitrary parameters to components") but more specifically about `CaptureUnmatchedValues`.

If we want to support passing unmatched parameters into this API, then we have to serialize out a descriptor of the type, otherwise there's no way to deserialize it later. The information doesn't exist anywhere else.

I don't think we want to impose the limitation that you can't pass unmatched parameters into the component rendered from here.

We *could* do something complicated whereby we only emit the type descriptor for unmatched parameters, but I'd say that's more of a possible (non-breaking) future refinement than a necessary feature for 3.1.

Reply:
> If we want to support passing unmatched parameters into this API, then we have to serialize out a descriptor of the type, otherwise there's no way to deserialize it later. The information doesn't exist anywhere else.

So that makes a lot of sense. I'm concerned about the impact of this on the 32kb message size limit.

Reply:
We need to understand if users will be able to use this feature in practical scenarios.